distributed/scheduler.py

```diff
@@ -1826,7 +1817,57 @@ def new_task(self, key: str, spec: object, state: str) -> TaskState:
     # State Transitions #
     #####################
 
-    def _transition(self, key, finish: str, *args, **kwargs):
+    def _transition_dispatch(self, start: str, finish: str, *args, **kwargs):
+        start = sys_intern(start)
+        finish = sys_intern(finish)
+
+        if start is "released":  # noqa: F632
+            if finish is "waiting":  # noqa: F632
+                return self.transition_released_waiting(*args, **kwargs)
+            elif finish is "forgotten":  # noqa: F632
+                return self.transition_released_forgotten(*args, **kwargs)
+            elif finish is "erred":  # noqa: F632
+                return self.transition_released_erred(*args, **kwargs)
+        elif start is "waiting":  # noqa: F632
+            if finish is "released":  # noqa: F632
+                return self.transition_waiting_released(*args, **kwargs)
+            elif finish is "processing":  # noqa: F632
+                return self.transition_waiting_processing(*args, **kwargs)
+            elif finish is "memory":  # noqa: F632
+                return self.transition_waiting_memory(*args, **kwargs)
+        elif start is "processing":  # noqa: F632
+            if finish is "released":  # noqa: F632
+                return self.transition_processing_released(*args, **kwargs)
+            elif finish is "memory":  # noqa: F632
+                return self.transition_processing_memory(*args, **kwargs)
+            elif finish is "erred":  # noqa: F632
+                return self.transition_processing_erred(*args, **kwargs)
+        elif start is "no-worker":  # noqa: F632
+            if finish is "released":  # noqa: F632
+                return self.transition_no_worker_released(*args, **kwargs)
+            elif finish is "waiting":  # noqa: F632
+                return self.transition_no_worker_waiting(*args, **kwargs)
+        elif start is "memory":  # noqa: F632
+            if finish is "released":  # noqa: F632
+                return self.transition_memory_released(*args, **kwargs)
+            elif finish is "forgotten":  # noqa: F632
+                return self.transition_memory_forgotten(*args, **kwargs)
+        elif start is "erred":  # noqa: F632
+            if finish is "released":  # noqa: F632
+                return self.transition_erred_released(*args, **kwargs)
+            elif finish is "forgotten":  # noqa: F632
+                return self.transition_released_forgotten(*args, **kwargs)
```

Comment:
From @mrocklin's comment ( https://github.com/dask/distributed/commit/8ff10d95b319b82282b6dc8742d889b939952fe7#r49507747 ).

> Does this change result in a meaningful speedup? This seems significantly more complex to me. I agree that it's worth doing, but only if it has an impact. If possible I think that we should try to measure the impact of this PR before merging. Thoughts?

(pasted here as the comment wasn't in the PR)

Reply:
So the goal of this change is to try and cutdown on time spent calling the `transition_*` functions themselves by moving those calls to C. To benefit from adding C APIs to those functions we need to call them directly. Doing it indirectly will just go through Python so won't have any benefit. IOW something like this will be needed

We could flatten the `if`s by comparing a single `str` (maybe like `"{start} {finish}"`?) to make this a bit simpler/more readable. May play with this a bit. Am open to other suggestions to improve readability/maintainability here

It's worth noting that today this lives in a `dict` that is roughly as long as this code block

Reply:
Yeah, I understand why, from first principles, this is a good idea.  I'm curious what impact is has on benchmarks and performance.  

If this improves benchmarks by a percent or two then awesome, let's good for it.  If it's only shaving off a couple of nanoseconds then maybe we shouldn't do this.  

I think that right now our development flow is to make some changes, merge them in, and then see how benchmarks respond.  I'm suggesting that we may want to roll those benchmarks in earlier in the process so that we make changes, look at how they change benchmarks/profiling, and then based on those improvements we decide to merge.  

Thoughts?

Reply:
Yeah I think that is fine. Was just trying to provide more context as well as make sure we are comfortable with this kind of change before diving deeper. Sound like we are ok with it in principle, but would like to make sure it is delivering value, which makes sense to me

Reply:
@jakirkham I've noticed you active on github again.  Welcome back from vacation?  I hope that you had a relaxing time.

I'm curious to learn how much impact this PR had.  I suspect that this is already on your queue of work coming back (which I imagine is long) but I thought I'd throw a :+1: on it anyway.

Reply:
Yeah I had another idea on how to improve things here, but as guessed other things are at the forefront of the queue atm 😅